// @directus/api/dist/controllers/server.js
import async_handler_default from "../utils/async-handler.js";
import { SettingsService } from "../services/settings.js";
import { respond } from "../middleware/respond.js";
import { ServerService } from "../services/server.js";
import { SpecificationService } from "../services/specifications.js";
import "../services/index.js";
import { createAdmin } from "../utils/create-admin.js";
import { useEnv } from "@directus/env";
import { ErrorCode, ForbiddenError, RouteNotFoundError, isDirectusError } from "@directus/errors";
import { Router } from "express";
import { format } from "date-fns";

//#region src/controllers/server.ts
const router = Router();
const env = useEnv();
if (env["OPENAPI_ENABLED"] !== false) router.get("/specs/oas", async_handler_default(async (req, res, next) => {
	const service = new SpecificationService({
		accountability: req.accountability,
		schema: req.schema
	});
	res.locals["payload"] = await service.oas.generate(req.headers.host);
	return next();
}), respond);
if (env["GRAPHQL_INTROSPECTION"] !== false) router.get("/specs/graphql/:scope?", async_handler_default(async (req, res) => {
	const service = new SpecificationService({
		accountability: req.accountability,
		schema: req.schema
	});
	const serverService = new ServerService({
		accountability: req.accountability,
		schema: req.schema
	});
	const scope = req.params["scope"] || "items";
	if (["items", "system"].includes(scope) === false) throw new RouteNotFoundError({ path: req.path });
	const info = await serverService.serverInfo();
	const result = await service.graphql.generate(scope);
	const filename = info["project"].project_name + "_" + format(/* @__PURE__ */ new Date(), "yyyy-MM-dd") + ".graphql";
	res.attachment(filename);
	res.send(result);
}));
router.get("/info", async_handler_default(async (req, res, next) => {
	const data = await new ServerService({
		accountability: req.accountability,
		schema: req.schema
	}).serverInfo();
	res.locals["payload"] = { data };
	return next();
}), respond);
router.get("/health", async_handler_default(async (req, res, next) => {
	const data = await new ServerService({
		accountability: req.accountability,
		schema: req.schema
	}).health();
	res.setHeader("Content-Type", "application/health+json");
	if (data["status"] === "error") res.status(503);
	res.locals["payload"] = data;
	res.locals["cache"] = false;
	return next();
}), respond);
router.post("/setup", async_handler_default(async (req, _res, next) => {
	if (await new ServerService({ schema: req.schema }).isSetupCompleted()) throw new ForbiddenError();
	try {
		await createAdmin(req.schema, {
			email: req.body.project_owner,
			password: req.body.password,
			first_name: req.body.first_name,
			last_name: req.body.last_name
		});
		new SettingsService({ schema: req.schema }).setOwner(req.body);
	} catch (error) {
		if (isDirectusError(error, ErrorCode.Forbidden)) return next();
		throw error;
	}
	return next();
}), respond);
var server_default = router;

//#endregion
export { server_default as default };
