// @directus/api/dist/database/helpers/fn/dialects/postgres.js
import { FnHelper } from "../types.js";
import { InvalidQueryError } from "@directus/errors";
import { toPath } from "lodash-es";

//#region src/database/helpers/fn/dialects/postgres.ts
const parseLocaltime = (columnType) => {
	if (columnType === "timestamp") return ` AT TIME ZONE 'UTC'`;
	return "";
};
var FnHelperPostgres = class extends FnHelper {
	year(table, column, options) {
		return this.knex.raw(`EXTRACT(YEAR FROM ??.??${parseLocaltime(options?.type)})`, [table, column]);
	}
	month(table, column, options) {
		return this.knex.raw(`EXTRACT(MONTH FROM ??.??${parseLocaltime(options?.type)})`, [table, column]);
	}
	week(table, column, options) {
		return this.knex.raw(`EXTRACT(WEEK FROM ??.??${parseLocaltime(options?.type)})`, [table, column]);
	}
	day(table, column, options) {
		return this.knex.raw(`EXTRACT(DAY FROM ??.??${parseLocaltime(options?.type)})`, [table, column]);
	}
	weekday(table, column, options) {
		return this.knex.raw(`EXTRACT(DOW FROM ??.??${parseLocaltime(options?.type)})`, [table, column]);
	}
	hour(table, column, options) {
		return this.knex.raw(`EXTRACT(HOUR FROM ??.??${parseLocaltime(options?.type)})`, [table, column]);
	}
	minute(table, column, options) {
		return this.knex.raw(`EXTRACT(MINUTE FROM ??.??${parseLocaltime(options?.type)})`, [table, column]);
	}
	second(table, column, options) {
		return this.knex.raw(`EXTRACT(SECOND FROM ??.??${parseLocaltime(options?.type)})`, [table, column]);
	}
	count(table, column, options) {
		const collectionName = options?.originalCollectionName || table;
		const type = this.schema.collections?.[collectionName]?.fields?.[column]?.type ?? "unknown";
		if (type === "json") {
			const { dbType } = this.schema.collections[table].fields[column];
			return this.knex.raw(dbType === "jsonb" ? "jsonb_array_length(??.??)" : "json_array_length(??.??)", [table, column]);
		}
		if (type === "alias") return this._relationalCount(table, column, options);
		throw new Error(`Couldn't extract type from ${table}.${column}`);
	}
	parseJsonResult(value) {
		return value;
	}
	json(table, column, options) {
		const collectionName = options?.originalCollectionName || table;
		const fieldSchema = this.schema.collections?.[collectionName]?.fields?.[column];
		if (!fieldSchema || fieldSchema.type !== "json" || !options?.jsonPath) throw new InvalidQueryError({ reason: `${collectionName}.${column} is not a JSON field` });
		const { template, bindings } = buildPostgresJsonPath(options.jsonPath);
		const cast = fieldSchema.dbType === "jsonb" ? "jsonb" : "json";
		return this.knex.raw(`??::${cast}${template}`, [table + "." + column, ...bindings]);
	}
};
/**
* Build a parameterized PostgreSQL JSON path using -> operators.
* Returns a template string containing only operators and ? placeholders,
* plus a bindings array with the actual values.
*
* @example ".color" → { template: "->?", bindings: ["color"] }
* @example ".items[0].name" → { template: "->?->?->?", bindings: ["items", 0, "name"] }
*/
function buildPostgresJsonPath(path) {
	const parts = toPath(path.startsWith(".") ? path.slice(1) : path);
	let template = "";
	const bindings = [];
	for (let i = 0; i < parts.length; i++) {
		const num = Number(parts[i]);
		template += "->?";
		if (!isNaN(num) && num >= 0 && Number.isInteger(num)) bindings.push(num);
		else bindings.push(parts[i]);
	}
	return {
		template,
		bindings
	};
}

//#endregion
export { FnHelperPostgres, buildPostgresJsonPath };
