// @directus/api/dist/ai/tools/fields/index.js
import { clearSystemCache } from "../../../cache.js";
import database_default from "../../../database/index.js";
import { transaction } from "../../../utils/transaction.js";
import { shouldClearCache } from "../../../utils/should-clear-cache.js";
import { requireText } from "../../../utils/require-text.js";
import { defineTool } from "../define-tool.js";
import { getSchema } from "../../../utils/get-schema.js";
import { FieldsService } from "../../../services/fields.js";
import { FieldItemInputSchema, FieldItemValidateSchema, RawFieldItemInputSchema, RawFieldItemValidateSchema } from "../schema.js";
import { InvalidPayloadError } from "@directus/errors";
import { toArray } from "@directus/utils";
import { fileURLToPath } from "node:url";
import { z } from "zod";
import { dirname, resolve } from "node:path";

//#region src/ai/tools/fields/index.ts
const __dirname = dirname(fileURLToPath(import.meta.url));
const FieldsBaseValidateSchema = z.strictObject({ collection: z.string() });
const FieldsValidateSchema = z.discriminatedUnion("action", [
	FieldsBaseValidateSchema.extend({
		action: z.literal("create"),
		data: z.union([z.array(FieldItemValidateSchema), FieldItemValidateSchema])
	}),
	z.object({
		action: z.literal("read"),
		collection: z.string().optional(),
		field: z.string().optional()
	}),
	FieldsBaseValidateSchema.extend({
		action: z.literal("update"),
		data: z.array(RawFieldItemValidateSchema.partial({ type: true }))
	}),
	FieldsBaseValidateSchema.extend({
		action: z.literal("delete"),
		field: z.string()
	})
]);
const FieldsInputSchema = z.object({
	action: z.enum([
		"read",
		"create",
		"update",
		"delete"
	]).describe("The operation to perform"),
	collection: z.string().describe("The name of the collection").optional(),
	field: z.string().describe("The name of the field. Required for delete. Optional for read (omit to read all fields). Do not use for create or update.").optional(),
	data: z.array(FieldItemInputSchema.extend({ children: RawFieldItemInputSchema.shape.children }).partial().required({ field: true })).describe("Array of field objects for create/update actions. Each object must include \"field\" (the field name).").optional()
});
const fields = defineTool({
	name: "fields",
	admin: true,
	description: requireText(resolve(__dirname, "./prompt.md")),
	annotations: { title: "Directus - Fields" },
	inputSchema: FieldsInputSchema,
	validateSchema: FieldsValidateSchema,
	async handler({ args, schema, accountability }) {
		let service = new FieldsService({
			schema,
			accountability
		});
		if (args.action === "create") {
			const fields$1 = toArray(args.data);
			const knex = database_default();
			const result = [];
			await transaction(knex, async (trx) => {
				service = new FieldsService({
					schema,
					accountability,
					knex: trx
				});
				for (const field of fields$1) await service.createField(args.collection, field, void 0, {
					autoPurgeCache: false,
					autoPurgeSystemCache: false
				});
			});
			if (shouldClearCache(service.cache)) await service.cache.clear();
			await clearSystemCache();
			service = new FieldsService({
				schema: await getSchema(),
				accountability
			});
			for (const field of fields$1) {
				const createdField = await service.readOne(args.collection, field.field);
				result.push(createdField);
			}
			return {
				type: "text",
				data: result || null
			};
		}
		if (args.action === "read") {
			let result = null;
			if (args.collection) if (args.field) result = await service.readOne(args.collection, args.field);
			else result = await service.readAll(args.collection);
			else result = await service.readAll();
			return {
				type: "text",
				data: result || null
			};
		}
		if (args.action === "update") {
			const fields$1 = toArray(args.data);
			const knex = database_default();
			const result = [];
			await transaction(knex, async (trx) => {
				service = new FieldsService({
					schema,
					accountability,
					knex: trx
				});
				for (const field of fields$1) await service.updateField(args.collection, field, {
					autoPurgeCache: false,
					autoPurgeSystemCache: false
				});
			});
			if (shouldClearCache(service.cache)) await service.cache.clear();
			await clearSystemCache();
			service = new FieldsService({
				schema: await getSchema(),
				accountability
			});
			for (const field of fields$1) {
				const updatedField = await service.readOne(args.collection, field.field);
				result.push(updatedField);
			}
			return {
				type: "text",
				data: result || null
			};
		}
		if (args.action === "delete") {
			const { collection, field } = args;
			await service.deleteField(collection, field);
			return {
				type: "text",
				data: {
					collection,
					field
				}
			};
		}
		throw new InvalidPayloadError({ reason: "Invalid action" });
	}
});

//#endregion
export { FieldsBaseValidateSchema, FieldsInputSchema, FieldsValidateSchema, fields };
