// @directus/api/dist/utils/versioning/handle-version.js
import { transaction } from "../transaction.js";
import { splitRecursive } from "./split-recursive.js";
import { ForbiddenError } from "@directus/errors";
import { deepMapWithSchema } from "@directus/utils";

//#region src/utils/versioning/handle-version.ts
async function handleVersion(self, key, queryWithKey, opts) {
	const { VersionsService } = await import("../../services/versions.js");
	const { ItemsService } = await import("../../services/items.js");
	if (queryWithKey.versionRaw) {
		const originalData = await self.readByQuery(queryWithKey, opts);
		if (originalData.length === 0) throw new ForbiddenError();
		const version$1 = await new VersionsService({
			schema: self.schema,
			accountability: self.accountability,
			knex: self.knex
		}).getVersionSave(queryWithKey.version, self.collection, key);
		return Object.assign(originalData[0], version$1?.delta);
	}
	let result;
	const versionsService = new VersionsService({
		schema: self.schema,
		accountability: self.accountability,
		knex: self.knex
	});
	const createdIDs = {};
	const version = await versionsService.getVersionSave(queryWithKey.version, self.collection, key, false);
	if (!version) throw new ForbiddenError();
	const { delta } = version;
	await transaction(self.knex, async (trx) => {
		const itemsServiceAdmin = new ItemsService(self.collection, {
			schema: self.schema,
			accountability: { admin: true },
			knex: trx
		});
		if (delta) {
			const { rawDelta, defaultOverwrites } = splitRecursive(delta);
			await itemsServiceAdmin.updateOne(key, rawDelta, {
				emitEvents: false,
				autoPurgeCache: false,
				skipTracking: true,
				overwriteDefaults: defaultOverwrites,
				onItemCreate: (collection, pk) => {
					if (collection in createdIDs === false) createdIDs[collection] = [];
					createdIDs[collection].push(pk);
				}
			});
		}
		result = (await new ItemsService(self.collection, {
			schema: self.schema,
			accountability: self.accountability,
			knex: trx
		}).readByQuery(queryWithKey, opts))[0];
		await trx.rollback();
	});
	if (!result) throw new ForbiddenError();
	return deepMapWithSchema(result, ([key$1, value], context) => {
		if (context.relationType === "m2o" || context.relationType === "a2o") {
			if (createdIDs[context.relation.related_collection]?.find((id) => String(id) === String(value))) return [key$1, null];
		} else if (context.relationType === "o2m" && Array.isArray(value)) {
			const ids = createdIDs[context.relation.collection];
			return [key$1, value.map((val) => {
				if (ids?.find((id) => String(id) === String(val))) return null;
				return val;
			})];
		}
		if (context.field.field === context.collection.primary) {
			if (createdIDs[context.collection.collection]?.find((id) => String(id) === String(value))) return [key$1, null];
		}
		return [key$1, value];
	}, {
		collection: self.collection,
		schema: self.schema
	});
}

//#endregion
export { handleVersion };
